NEW Added tests for existing math operators

jkrude opened the pull request:
Rework of #828
Tests were run on a cuda-device and std-backend.
See also https://github.com/ComputationalRadiationPhysics/alpaka/pull/828 for the previous version.
Implemented functors and abstraction, so that one code is used for all operators.

tdd11235813 commented:
thanks @jkrude. In our case we took a more involved code to keep the definition and extensions of math operations compact and to learn C++11. I am happy, that you use patterns of meta-programming, so you are getting into Alpaka :) You also document the code, that's good!

However, the code is too complex and error-prone to leave it untested itself, e.g. how do I know, that the arguments' data are filled correctly?
I am going to prepare a PR for your side, since there are several things that need be fixed in the interface and coding style.

BenjaminW3 commented:
Can #828 be closed now?

tdd11235813 commented:
@jkrude is preparing an update to this PR next days. I gonna close the #828.

sbastrakov commented on test/unit/math/mathOps/include/Defines.hpp:
```diff
@@ -0,0 +1,39 @@
+    template< typename T >
+    auto rsqrt( T const & arg ) -> decltype( std::sqrt( arg ) )
+    {
+        return 1 / std::sqrt( arg );
```
Should not this `1` be `static_cast`ed to `T`?

sbastrakov commented on test/unit/math/mathOps/CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright 2017-2019 Benjamin Worpitz, Axel Huebl, Matthias Werner
+# Copyright 2017-2019 Benjamin Worpitz, Jakob Krude
```
I assume this is accidental? 
(We never remove people from the copyright line, only add, unless originally added by mistake)

jkrude replied:
It was a copy-paste mistake in the beginning.

sbastrakov replied:
Ah, thanks for clarifying.

psychocoderHPC commented:
~~Do not wonder I stopped the CI because we need to prioritize another PR.  I will start the test by hand later.~~
Restarted the tests now.

BenjaminW3 commented:
There are ome compilation errors when clang is used as native CUDA compiler:
```
In file included from /home/travis/build/ComputationalRadiationPhysics/alpaka/test/unit/math/mathOps/src/mathOps.cpp:12:
/home/travis/build/ComputationalRadiationPhysics/alpaka/test/unit/math/mathOps/src/../include/Functor.hpp:120:1: error: reference to __host__ function 'execute<nullptr_t, double, 0>' in __host__ __device__ function
ALPAKA_TEST_MATH_OP_FUNCTOR( OpAbs,
^
/home/travis/build/ComputationalRadiationPhysics/alpaka/test/unit/math/mathOps/src/../include/Functor.hpp:90:17: note: expanded from macro 'ALPAKA_TEST_MATH_OP_FUNCTOR'
    -> decltype(execute(acc, args.arg[0]))                                      \
                ^
/home/travis/build/ComputationalRadiationPhysics/alpaka/test/unit/math/mathOps/src/../include/Functor.hpp:120:1: note: 'execute<nullptr_t, double, 0>' declared here
/home/travis/build/ComputationalRadiationPhysics/alpaka/test/unit/math/mathOps/src/../include/Functor.hpp:72:10: note: expanded from macro 'ALPAKA_TEST_MATH_OP_FUNCTOR'
    auto execute(                                                               \
         ^
/home/travis/build/ComputationalRadiationPhysics/alpaka/test/unit/math/mathOps/src/mathOps.cpp:40:29: note: called by 'operator()<alpaka::acc::AccGpuCudaRt<std::integral_constant<unsigned long, 1>, unsigned long>, alpaka::test::unit::math::Buffer<alpaka::acc::AccGpuCudaRt<std::integral_constant<unsigned long, 1>, unsigned long>, double, 1000>, alpaka::test::unit::math::OpAbs, alpaka::test::unit::math::Buffer<alpaka::acc::AccGpuCudaRt<std::integral_constant<unsigned long, 1>, unsigned long>, alpaka::test::unit::math::ArgsItem<double, alpaka::test::unit::math::Arity::UNARY>, 1000> >'
          results(i, acc) = functor(args(i, acc), acc);
                            ^
/home/travis/build/ComputationalRadiationPhysics/alpaka/include/alpaka/kernel/TaskKernelGpuCudaRt.hpp:88:21: note: called by 'cudaKernel<std::integral_constant<unsigned long, 1>, unsigned long, TestKernel, alpaka::test::unit::math::Buffer<alpaka::acc::AccGpuCudaRt<std::integral_constant<unsigned long, 1>, unsigned long>, double, 1000>, alpaka::test::unit::math::OpAbs, alpaka::test::unit::math::Buffer<alpaka::acc::AccGpuCudaRt<std::integral_constant<unsigned long, 1>, unsigned long>, alpaka::test::unit::math::ArgsItem<double, alpaka::test::unit::math::Arity::UNARY>, 1000> >'
                    kernelFnObj(
                    ^
```

BenjaminW3 commented on test/unit/math/mathOps/include/Functor.hpp:
```diff
@@ -0,0 +1,298 @@
+    alpaka::math::rsqrt,
+    Range::POSITIVE_ONLY )
+
+    ALPAKA_TEST_MATH_OP_FUNCTOR( OpSin,
```
Too many leading spaces here.

BenjaminW3 commented on test/unit/math/mathOps/include/Functor.hpp:
```diff
@@ -0,0 +1,298 @@
+    //! @param ARITY Enum-type can be one ... n
+    //! @param STD_OP Function used for the host side, e.g. std::abs
+    //! @param ALPAKA_OP Function used for the device side, e.g. alpaka::math::abs.
+    //! @param ... List of Ranges. Needs to match the arity.
```
Can you please remove the leading spaces here?

BenjaminW3 commented on test/unit/math/mathOps/include/Defines.hpp:
```diff
@@ -0,0 +1,72 @@
+    enum class Arity
+    {
+        UNARY = 1,
+        BINARY = 2
```
For consistency with the other `enum`s in alpaka (e.g. `enum class GridBlockExtentSubDivRestrictions`) we should make the values `CamelCase`.

tdd11235813 replied:
also for `enum class Range`

ax3l commented on test/unit/math/mathOps/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+    ${_TARGET_NAME}
+    PROPERTIES
+    COMPILE_OPTIONS
+    $<$<CXX_COMPILER_ID:clang>:"-ffp-contract=off"> # ffp-contract: https://llvm.org/docs/CompileCudaWithLLVM.html#id5
```
this disables fused multiply-adds altogether. why do we want this? It's exactly the opposite to what we do above for nvcc/hip...

Also, the anchor in the link should read https://llvm.org/docs/CompileCudaWithLLVM.html#flags-that-control-numerical-code in order to work.

tdd11235813 replied:
was my bad, just read the `ffast-math` line and assumed that the flag defaults have to be disabled, while we actually only need `-fcuda-approx-transcendentals` to be off, because sin/cos will return mismatches for large argument values when using fast math settings.

ax3l replied:
no prob, that's why we do code review :)

psychocoderHPC replied:
@jkrude could you please remove this line

BenjaminW3 commented:
Is anyone working on fixing the compilation issues?

psychocoderHPC commented:
@jkrude  PLease rebase this PR against the current dev and squash all commits.
I tested this PR with HIP-clang 3.1 on our dev system fwk388 and all works fine.

jkrude committed:
Added unit test for n-arity math operators

Added tests for unary math operators with functors.

Added binaryOps and reworked unaryOps.

Reduced doubled code significantly, by allowing to test operators of n-arity with the same code.

Integrated PR from td11235813

Removed HCC workaround and rebased
NOTE: fast-math needs to be disabled for gpu-backend.
Co-authored-by: Matthias Werner <Matthias.Werner1@tu-dresden.de>

BenjaminW3 commented on test/unit/math/mathOps/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+
+SET_TARGET_PROPERTIES(${_TARGET_NAME} PROPERTIES FOLDER "test/unit")
+
+ADD_TEST(NAME ${_TARGET_NAME} COMMAND ${_TARGET_NAME} ${_ALPAKA_TEST_OPTIONS})
```
Could you please make the cmake commands (but not the variables) in this file lowercase to match the changed style in all other files?

BenjaminW3 commented on test/unit/CMakeLists.txt:
```diff
@@ -18,20 +18,21 @@ cmake_minimum_required(VERSION 3.15)
+ADD_SUBDIRECTORY("queue/")
+ADD_SUBDIRECTORY("rand/")
+ADD_SUBDIRECTORY("time/")
+ADD_SUBDIRECTORY("vec/")
```
here lowercase as well

BenjaminW3 commented on test/unit/math/mathOps/include/DataGen.hpp:
```diff
@@ -0,0 +1,139 @@
+            "Set of args must provide > 2 entries." );
+        constexpr auto max = std::numeric_limits< TData >::max();
+        constexpr auto low = std::numeric_limits< TData >::lowest();
+        std::default_random_engine eng{ seed };
```
You get a cast warning here:
`error: non-constant-expression cannot be narrowed from type 'unsigned long' to 'std::__1::linear_congruential_engine<unsigned int, 48271, 0, 2147483647>::result_type' (aka 'unsigned int') in initializer list [-Wc++11-narrowing]`
This can probably be fixed by:
```suggestion
        std::default_random_engine eng{ static_cast<std::default_random_engine::result_type>(seed) };
```

BenjaminW3 commented on test/unit/math/mathOps/include/Functor.hpp:
```diff
@@ -0,0 +1,288 @@
+    auto operator()(                                                            \
+        ArgsItem<T, Arity::Binary> const & args,                                \
+        TAcc const & acc = nullptr) const                                       \
+    -> decltype(execute(acc, args.arg[0], args.arg[1]))                         \
```
Now that we have C++14 as baseline, we can remove this line (here and in all others above).

psychocoderHPC replied:
It could be that the return type is required because of CUDA restrictions. We need to test it here.

BenjaminW3 commented on test/unit/math/mathOps/include/Functor.hpp:
```diff
@@ -0,0 +1,288 @@
+             typename std::enable_if<                                           \
+                 std::is_same< TAcc, std::nullptr_t>::value,                    \
+                 int>::type = 0>                                                \
+    ALPAKA_FN_HOST                                                              \
```
clang as CUDA compiler complains: `error: reference to __host__ function 'execute<nullptr_t, double, 0>' in __host__ __device__ function`

jkrude committed:
implemented suggested PR-changes

BenjaminW3 commented:
Some of the windows builds seem to have problems like:
`C:\Users\RUNNER~1\AppData\Local\Temp\tmpxft_000007f8_00000000-9_mathOps.cudafe1.cpp : fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj [D:\a\alpaka\alpaka\build\test\unit\math\mathOps\mathOps.vcxproj]`
You can solve this by adding:
```
target_compile_options(${_COMMON_TARGET_NAME} PUBLIC "/bigobj")
```
after line 37 in `alpaka/test/common/CMakeLists.txt`.

BenjaminW3 commented on test/unit/math/mathOps/src/mathOps.cpp:
```diff
@@ -0,0 +1,253 @@
+        TAcc const & acc,
+        TResults const & results,
+        TFunctor const & functor,
+        TArgs const & args) const
```
gcc 6 and 7 complain about
```
/home/travis/build/alpaka-group/alpaka/test/unit/math/mathOps/src/mathOps.cpp:31:24: error: but 'void TestKernel::operator()(const TAcc&, const TResults&, const TFunctor&, const TArgs&) const [with TAcc = alpaka::acc::AccCpuSerial<std::integral_constant<long unsigned int, 1ul>, long unsigned int>; TResults = alpaka::test::unit::math::Buffer<alpaka::acc::AccCpuSerial<std::integral_constant<long unsigned int, 1ul>, long unsigned int>, double, 1000ul>; TFunctor = alpaka::test::unit::math::OpAcos; TArgs = alpaka::test::unit::math::Buffer<alpaka::acc::AccCpuSerial<std::integral_constant<long unsigned int, 1ul>, long unsigned int>, alpaka::test::unit::math::ArgsItem<double, (alpaka::test::unit::math::Arity)1>, 1000ul>]' does not throw; perhaps it should be declared 'noexcept' [-Werror=noexcept]
```
We could try adding `noexcept`

BenjaminW3 commented:
I created a PR to fix the Windows bigobj issue: #979

sbastrakov commented on test/unit/math/mathOps/src/mathOps.cpp:
```diff
@@ -0,0 +1,254 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "../include/Defines.hpp"
+#include "../include/Buffer.hpp"
+#include "../include/Functor.hpp"
+#include "../include/DataGen.hpp"
```
Sorry, I did not notice this in time. I believe it's better to avoid relative paths here, and also in other files added by this PR, and instead do `#include <alpaka/test/unit/math/mathOps/include/Defines.hpp>`. What do you think @jkrude @BenjaminW3 @psychocoderHPC ?

BenjaminW3 replied:
The CMake way would be to add the include folder to the `target_include_directories` of the test target. Then you do not need relative includes anymore.

BenjaminW3 replied:
But on the other hand those files are no public includes at all. We should simply move them into the `src` folder.

psychocoderHPC replied:
Moving them into `src` is IMO ok. The current relative path requires that the install folder structure  is equal to the structure we have in our source code. We should not assume it.

sbastrakov replied:
@jkrude when you have time, could you implement this change? Should be not much work, just moving the files and fixing the includes.

psychocoderHPC replied:
> @jkrude when you have time, could you implement this change? Should be not much work, just moving the files and fixing the includes.

Already fixed and merged with #986
